<a href="https://colab.research.google.com/github/MonicaMaVi/Algortimos_de_optimizacion/blob/master/Seminario_Algoritmos.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Algoritmos de optimización - Seminario<br>
Nombre y Apellidos: Mónica Mayayo Vives  <br>
Url: github.com/MonicaMaVi/Algortimos_de_optimizacion/Seminario_Algoritmos.ipynb<br>
Problema: 3. Combinar cifras y operaciones

Problema 3. Combinar cifras y operaciones
• El problema consiste en analizar el siguiente problema y diseñar un algoritmo que lo resuelva.
• Disponemos de las 9 cifras del 1 al 9 (excluimos el cero) y de los 4 signos básicos de las
operaciones fundamentales: suma(+), resta(-), multiplicación(*) y división(/)
• Debemos combinarlos alternativamente sin repetir ninguno de ellos para obtener una
cantidad dada. Un ejemplo sería para obtener el 4:
4+2-6/3*1 = 4
• Debe analizarse el problema para encontrar todos los valores enteros posibles planteando las
siguientes cuestiones:
- ¿Qué valor máximo y mínimo se pueden obtener según las condiciones del problema?

Se puede obtener un mínimo negativo ya que tenemos la operación menos y se pueden obtener números bastante alejados del 0, tanto negativos como positivos ya que tenemos la operación de multiplicación. En el código de abajo se adjunta el cálculo del máximo y el mínimo.

- ¿Es posible encontrar todos los valores enteros posibles entre dicho mínimo y máximo?

En el rango de enteros [-69, 77] no hay ningún número entero que no aparezca, si probamos con los números que están en los extremos, como -69, -68... y 77, 76... (cambiando el 77 en el código adjuntado abajo por el número que queramos probar) aparecen, por lo que podemos intuir que cuanto más nos acercamos al centro del rango (probando con 0, 1, -1) más veces van a aparecer esos números como resultado, y probándolo en el código así es. Es más probable que más operaciones den un resultado más próximo al centro del rango, en este caso el cuatro, al tener operaciones de suma y resta, por lo que si los números del extremo están, se intuye que los números centrales también estarán. 

He calculado en python, en el código abajo adjuntado, los números enteros que aparecen entre el máximo y el mínimo, los he guardado en un vector, he quitado los repetidos, he ordenado el vector y lo he mostrado en el resultado del programa, por lo que se puede ver que efectivamente aparecen todos los números enteros entre el máximo y el mínimo.

Si calculáramos el mínimo y el máximo cogiendo números decimales (nos daría el mínimo un valor de -70.71 y el máximo un valor de 78.83), allí sí que encontramos que el 78 que está entre el máximo y el mínimo pero no aparecen como resultado de ninguna fórmula obtenida.

In [ ]:
import itertools

N=list(range(1, 10))
signos = ['+', '-', '*', '/']
enteros=[]
minimo = 0
maximo = 0

# Permutaciones sin repetición de los números del 1 al 9 cogidos del 5 en 5
permuNum = list(itertools.permutations(N, 5))
# Permutaciones sin repetición de todos los signos
permuSignos = list(itertools.permutations(signos))

# Calculamos todas las fórmulas posibles
for i in permuNum: 
    for j in permuSignos: 
        formula = f"{i[0]}{j[0]}{i[1]}{j[1]}{i[2]}{j[2]}{i[3]}{j[3]}{i[4]}"
        resultado = eval(formula)
        # Cogemos solo números enteros
        if resultado==int(resultado):
            # vamos a crear un vector de resultados enteros para encontrar 
            # todos los valores enteros posibles entre el mínimo y el máximo.
            enteros.append(resultado)
            sin_repetidos = list(set(enteros)) # Quitamos los repetidos
            sin_repetidos.sort() # Lo ordenamos
            #Aquí mostraremos los resultados y las fórmulas que sean igual al número que elijamos, en este caso el 77
            if resultado == 77:
                print(formula)
                print(resultado)
            minimo = min(resultado,minimo)
            maximo = max(resultado,maximo)   

# Opcional: imprimir la cantidad total de permutaciones
print(f"Total de permutaciones: {len(permuNum)*len(permuSignos)}")
# Mostramos mínimo y máximo
print("Mínimo:", minimo, "Máximo:", maximo)
# Mostramos valores enteros entre el mínimo y el máximo
print("Valores enteros entre el mínimo y el máximo", sin_repetidos)

7/1-2+8*9
77.0
7/1-2+9*8
77.0
7/1+8*9-2
77.0
7/1+9*8-2
77.0
7-2/1+8*9
77.0
7-2/1+9*8
77.0
7-2+8/1*9
77.0
7-2+8*9/1
77.0
7-2+9/1*8
77.0
7-2+9*8/1
77.0
7-4/2+8*9
77.0
7-4/2+9*8
77.0
7-6/3+8*9
77.0
7-6/3+9*8
77.0
7+8/1*9-2
77.0
7+8*9/1-2
77.0
7+8*9-2/1
77.0
7+8*9-4/2
77.0
7+8*9-6/3
77.0
7+9/1*8-2
77.0
7+9*8/1-2
77.0
7+9*8-2/1
77.0
7+9*8-4/2
77.0
7+9*8-6/3
77.0
8/1*9-2+7
77.0
8/1*9+7-2
77.0
8*9/1-2+7
77.0
8*9/1+7-2
77.0
8*9-2/1+7
77.0
8*9-2+7/1
77.0
8*9-4/2+7
77.0
8*9-6/3+7
77.0
8*9+7/1-2
77.0
8*9+7-2/1
77.0
8*9+7-4/2
77.0
8*9+7-6/3
77.0
9/1*8-2+7
77.0
9/1*8+7-2
77.0
9*8/1-2+7
77.0
9*8/1+7-2
77.0
9*8-2/1+7
77.0
9*8-2+7/1
77.0
9*8-4/2+7
77.0
9*8-6/3+7
77.0
9*8+7/1-2
77.0
9*8+7-2/1
77.0
9*8+7-4/2
77.0
9*8+7-6/3
77.0
Total de permutaciones: 362880
Mínimo: -69.0 Máximo: 77.0
Valores enteros entre el mínimo y el máximo [-69.0, -68.0, -67.0, -66.0, -65.0, -64.0, -63.0, -62.0, -61.0, -60.0, -59.0, -58.0, -57.0, -56.0, -55.0, -54.0, -53.0, -52.0, -51.0, -50.0, -49.0, -48.0, -47.0, -46.0, -45.0, -4

(*)¿Cuantas posibilidades hay sin tener en cuenta las restricciones?<br>

¿Cuantas posibilidades hay teniendo en cuenta todas las restricciones.


(*)¿Cuantas posibilidades hay sin tener en cuenta las restricciones?<br>

El número total de posibilidades son las permutaciones de los números (15120) multiplicadas por las permutaciones de los signos (24) que dan un total de 362880. Las permutaciones de los nueve números cogidos de cinco en cinco nos dan una cantidad de 15120 ya que nos quedaría una operación de factorial de 9 en el numerador y en el denominador (9-5)!. Las permutaciones de los cuatro signos cogidos de cuatro en cuatro son 24 posibilidades, ya que sería 4! en el numerador y en el denominador (4-4)! = 0! que es 1. Por último multiplicaríamos 15120*24 lo que nos da los 362880 números totales.

¿Cuantas posibilidades hay teniendo en cuenta todas las restricciones?

Teniendo encuenta la restricción de que solo cogemos números enteros, sería difícil calcularlo matemáticamente, por lo que esa operación la he hecho programando el ejercicio en python (que adjunto abajo) y me da un total de 90000 posibilidades. Otra restricción que plantea el ejercicio sería mirar que la suma de esas fórmulas me dé un cierto número, por ejemplo el 4, también lo he calculado en el ejercicio en python que adjunto, se podría calcular para cualquier número cambiando en el código el 4 por cualquier otro número. 

In [6]:
import itertools

N=list(range(1, 10))
signos = ['+', '-', '*', '/']
cantidad = 0
cantidad4 = 0

# Permutaciones sin repetición de los números del 1 al 9 cogidos del 5 en 5
permuNum = list(itertools.permutations(N, 5))
# Permutaciones sin repetición de todos los signos
permuSignos = list(itertools.permutations(signos))

# Calculamos todas las fórmulas posibles
for i in permuNum: 
    for j in permuSignos: 
        formula = f"{i[0]}{j[0]}{i[1]}{j[1]}{i[2]}{j[2]}{i[3]}{j[3]}{i[4]}"
        resultado = eval(formula)
        # Cogemos solo números enteros
        if resultado==int(resultado):
            cantidad +=1 
            #Aquí mostraremos los resultados y las fórmulas que sean igual al número que elijamos, en este caso el 4
            if resultado == 4:
                cantidad4 +=1

# Opcional: imprimir la cantidad total de permutaciones
print(f"Número total de permutaciones: {len(permuNum)*len(permuSignos)}")
print(f"Número total de permutaciones que son números enteros:", cantidad)
print(f"Número total de permutaciones que suman 4:", cantidad4)


Número total de permutaciones: 362880
Número total de permutaciones que son números enteros: 90000
Número total de permutaciones que suman 4: 2112


Modelo para el espacio de soluciones<br>
(*) ¿Cual es la estructura de datos que mejor se adapta al problema? Argumentalo.(Es posible que hayas elegido una al principio y veas la necesidad de cambiar, arguentalo)


La estructura de datos que mejor se adapta es un f-string (formatted string) para poder manejarlo con la función eval, al principio elegí un vector para manejar la información de las fórmulas pero pronto me dí cuenta que no me servía para poder transformarlo en una fórmula operable con eval.

In [37]:
formula = f"{i[0]}{j[0]}{i[1]}{j[1]}{i[2]}{j[2]}{i[3]}{j[3]}{i[4]}"

Según el modelo para el espacio de soluciones<br>
(*)¿Cual es la función objetivo?

(*)¿Es un problema de maximización o minimización?

La función objetivo que se ha marcado en el enunciado es calcular el máximo y el mínimo de los resultados de las distintas fórmulas que se van creando, por lo que tenemos un problema tanto de maximización como de minimización, ya que podemos escribir la fórmula como una función, que en nuestro caso las incógnitas no serían solo los dígitos sino también los signos, y tenemos que encontrar el máximo y el mínimo de esa función. Adjunto abajo el código del cálculo del máximo y el mínimo.

In [39]:
import itertools

N=list(range(1, 10))
signos = ['+', '-', '*', '/']
minimo = 0
maximo = 0

# Permutaciones sin repetición de los números del 1 al 9 cogidos del 5 en 5
permuNum = list(itertools.permutations(N, 5))
# Permutaciones sin repetición de todos los signos
permuSignos = list(itertools.permutations(signos))

# Calculamos todas las fórmulas posibles
for i in permuNum: 
    for j in permuSignos: 
        formula = f"{i[0]}{j[0]}{i[1]}{j[1]}{i[2]}{j[2]}{i[3]}{j[3]}{i[4]}"
        resultado = eval(formula)
        # Cogemos solo números enteros
        if resultado==int(resultado):
            minimo = min(resultado,minimo)
            maximo = max(resultado,maximo)   

# Mostramos el mínimo y máximo
print("Mínimo:", minimo, "Máximo:", maximo)

Mínimo: -69.0 Máximo: 77.0


Diseña un algoritmo para resolver el problema por fuerza bruta

(Código adjuntado abajo) Para el problema por fuerza bruta generamos primero 2 listas de tuplas, una de números, donde las tuplas serán de longitud 5 y otra de signos, que será un listado de tuplas de longitud 4. El problema se puede realizar con 2 for, uno para recorrer los números y otro para recorrer los signos, dentro de estos bucles iremos comprobando si el nuevo número entero calculado es menor que el mínimo o mayor que el máximo en cuyo caso lo estableceremos como mínimo o máximo, según corresponda.

In [ ]:
import itertools

N=list(range(1, 10))
signos = ['+', '-', '*', '/']
minimo = 0
maximo = 0

# Permutaciones sin repetición de los números del 1 al 9 cogidos del 5 en 5
permuNum = list(itertools.permutations(N, 5))
# Permutaciones sin repetición de todos los signos
permuSignos = list(itertools.permutations(signos))

# Calculamos todas las fórmulas posibles
for i in permuNum: 
    for j in permuSignos: 
        formula = f"{i[0]}{j[0]}{i[1]}{j[1]}{i[2]}{j[2]}{i[3]}{j[3]}{i[4]}"
        resultado = eval(formula)
        # Cogemos solo números enteros
        if resultado==int(resultado):
            minimo = min(resultado,minimo)
            maximo = max(resultado,maximo)   

# Mostramos el mínimo y máximo
print("Mínimo:", minimo, "Máximo:", maximo)

Calcula la complejidad del algoritmo por fuerza bruta

Respuesta

(*)Diseña un algoritmo que mejore la complejidad del algortimo por fuerza bruta. Argumenta porque crees que mejora el algoritmo por fuerza bruta

Respuesta

(*)Calcula la complejidad del algoritmo

Respuesta

Según el problema (y tenga sentido), diseña un juego de datos de entrada aleatorios

Respuesta

Aplica el algoritmo al juego de datos generado

Respuesta

Enumera las referencias que has utilizado(si ha sido necesario) para llevar a cabo el trabajo

Respuesta

Describe brevemente las lineas de como crees que es posible avanzar en el estudio del problema. Ten en cuenta incluso posibles variaciones del problema y/o variaciones al alza del tamaño

Respuesta